lib/websocket/websocket.go

```diff
@@ -0,0 +1,321 @@
+package websocket
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+
+	"github.com/google/uuid"
+	golog "github.com/ipfs/go-log"
+	"github.com/qri-io/qri/auth/key"
+	"github.com/qri-io/qri/auth/token"
+	"github.com/qri-io/qri/event"
+	"nhooyr.io/websocket"
+	"nhooyr.io/websocket/wsjson"
+)
+
+const qriWebsocketProtocol = "qri-websocket"
+
+var (
+	errNotFound = fmt.Errorf("connection not found")
+
+	log = golog.Logger("websocket")
+)
+
+// newID returns a new websocket connection ID
+func newID() string {
+	return uuid.New().String()
+}
+
+// setIDRand sets the random reader that NewID uses as a source of random bytes
+// passing in nil will default to crypto.Rand. This can be used to make ID
+// generation deterministic for tests. eg:
+//    myString := "SomeRandomStringThatIsLong-SoYouCanCallItAsMuchAsNeeded..."
+//    lib.SetIDRand(strings.NewReader(myString))
+//    a := NewID()
+//    lib.SetIDRand(strings.NewReader(myString))
+//    b := NewID()
+func setIDRand(r io.Reader) {
+	uuid.SetRand(r)
+}
+
+// Handler defines the handler interface
+type Handler interface {
+	ConnectionHandler(w http.ResponseWriter, r *http.Request)
+}
+
+// connections maintains the set of active websocket connections & associated
+// connection metadata
+type connections struct {
+	conns     map[string]*conn
+	connsLock sync.Mutex
+	keystore  key.Store
+	// TODO(ramfox): using a `map[string]string` to track connections and
+	// profile.IDs means that each profile can only have one connection
+	// which will cause two browser tabs from the same profile to fail.
+	// we need to support multiple connections for the same profile, which
+	// will require an array of connection ID strings rather than a single
+	// string
+	subscriptions map[string]string
+	subsLock      sync.Mutex
+}
+
+type conn struct {
+	id        string
+	profileID string
+	conn      *websocket.Conn
+}
+
+var _ Handler = (*connections)(nil)
+
+// NewHandler creates a new connections instance that clients
+// can connect to in order to get realtime events
+func NewHandler(ctx context.Context, bus event.Bus, keystore key.Store) (Handler, error) {
+	ws := &connections{
+		conns:         map[string]*conn{},
+		connsLock:     sync.Mutex{},
+		keystore:      keystore,
+		subscriptions: map[string]string{},
+		subsLock:      sync.Mutex{},
+	}
+
+	bus.SubscribeAll(ws.messageHandler)
+	return ws, nil
+}
+
+// ConnectionHandler handles websocket upgrade requests and accepts the connection
+func (h *connections) ConnectionHandler(w http.ResponseWriter, r *http.Request) {
+	wsc, err := websocket.Accept(w, r, &websocket.AcceptOptions{
+		Subprotocols:       []string{qriWebsocketProtocol},
+		InsecureSkipVerify: true,
+	})
+	if err != nil {
+		log.Debugf("Websocket accept error: %s", err)
+		return
+	}
+	id := newID()
+	c := &conn{
+		id:   id,
+		conn: wsc,
+	}
+	h.connsLock.Lock()
+	defer h.connsLock.Unlock()
+	h.conns[id] = c
+	go h.read(r.Context(), id)
```

Comment:
hmmm, adding the request context here is causing the connection to be canceled very quickly. 

The examples in the `nhooyr/websocket` repo use the request context in this way, so I'm not sure what is different about our request context that is making this context cancel prematurely.

Reply:
hrm, ok. Let's file a follow-on issue after this is merged to look into it. Shouldn't be a show-stopping thing

Reply:
filed issue referencing this convo https://github.com/qri-io/qri/issues/1925